src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -0,0 +1,1011 @@
+            _outgoingBuffer.Consume(_outgoingBuffer.ReadMemory.Length);
+
+            _expectingSettingsAck = true;
+
+            // Receive the initial SETTINGS frame from the peer.
+            FrameHeader frameHeader = await ReadFrameAsync().ConfigureAwait(false);
+            if (frameHeader.Type != FrameType.Settings || frameHeader.AckFlag)
+            {
+                throw new Http2ProtocolException(Http2ProtocolErrorCode.ProtocolError);
+            }
+
+            // Process the SETTINGS frame.  This will send an ACK.
+            await ProcessSettingsFrame(frameHeader).ConfigureAwait(false);
+
+            ProcessIncomingFrames();
+        }
+
+        private int GetNewStreamId()
+        {
+            int newStream = _nextStream;
+            while (true)
+            {
+                if (newStream == MaxStreamId)
+                {
+                    // No more stream Ids available
+                    return -1;
+                }
+
+                int original = Interlocked.CompareExchange(ref _nextStream, newStream + 2, newStream);
+                if (original == newStream)
+                {
+                    return newStream;
+                }
+
+                newStream = original;
+            }
+        }
+
+        private async Task EnsureIncomingBytesAsync(int minReadBytes)
+        {
+            if (_incomingBuffer.ReadBytes.Length >= minReadBytes)
+            {
+                return;
+            }
+
+            int bytesNeeded = minReadBytes - _incomingBuffer.ReadBytes.Length;
+            _incomingBuffer.EnsureWriteSpace(bytesNeeded);
+            int bytesRead = await ReadAtLeastAsync(_stream, _incomingBuffer.WriteMemory, bytesNeeded).ConfigureAwait(false);
+            _incomingBuffer.Commit(bytesRead);
+        }
+
+        private async Task FlushOutgoingBytesAsync()
+        {
+            Debug.Assert(_outgoingBuffer.ReadBytes.Length > 0);
+
+            await SendFramesAsync(_outgoingBuffer.ReadMemory).ConfigureAwait(false);
+            _outgoingBuffer.Consume(_outgoingBuffer.ReadMemory.Length);
+        }
+
+        private async Task<FrameHeader> ReadFrameAsync()
+        {
+            // Read frame header
+            await EnsureIncomingBytesAsync(FrameHeader.Size).ConfigureAwait(false);
+            FrameHeader frameHeader = FrameHeader.Read(_incomingBuffer.ReadBytes);
+            _incomingBuffer.Consume(FrameHeader.Size);
+
+            if (frameHeader.Length > FrameHeader.MaxLength)
+            {
+                throw new Http2ProtocolException(Http2ProtocolErrorCode.FrameSizeError);
+            }
+
+            // Read frame contents
+            await EnsureIncomingBytesAsync(frameHeader.Length).ConfigureAwait(false);
+
+            return frameHeader;
+        }
+
+        private async void ProcessIncomingFrames()
+        {
+            try
+            {
+                while (true)
```

Comment:
So this is an asynchronous, infinite processing loop?  If one of those exceptions below occurs, it's going to get propagated to the sync ctx that was current when this method was called, which likely means it'll propagate on the thread pool and crash the process.

Reply:
(Nevermind my crashing comment... those throws are inside a catch block that then just eats them.)

Reply:
Yes, it's an infinite processing loop that should terminate when one of these things happens:
(1) We decide to close the connection ourselves (e.g. idle timeout or handler dispose)
(2) The server initiates a graceful shutdown by sending us a GOAWAY frame
(3) IO or protocol error on the connection, which will cause us to abort the connection

Only (3) is implemented currently, and even that not completely. I think we probably need to do something better here before merging, so I'll rework this.

Reply:
Note (2) does not mean the connection is finished, it means don't create any new streams. Existing streams are fine to keep processing until explicitly reset or the connection drops.

Reply:
Yes, that's how the GOAWAY handling works now. I pushed a new commit since the previous comment.